server/src/main/java/org/elasticsearch/common/breaker/PreallocatedCircuitBreakerService.java

```diff
@@ -0,0 +1,204 @@
+    }
+
+    @Override
+    public void close() {
+        preallocated.close();
+    }
+
+    /**
+     * The preallocated breaker.
+     * <p>
+     * This breaker operates in two states:
+     * <ol>
+     * <li>We've used fewer bytes than we've preallocated.
+     * <li>We've used all of the preallocated bytes.
+     * </ol>
+     * <p>
+     * If we're in the "used fewer bytes" state than we've allocated then
+     * allocating new bytes just adds to
+     * {@link PreallocedCircuitBreaker#preallocationUsed}, maxing out at
+     * {@link PreallocedCircuitBreaker#preallocated}. If we max
+     * out we irreversibly switch to "used all" state. In that state any
+     * additional allocations are passed directly to the underlying breaker.
+     * <p>
+     * De-allocating is just allocating a negative number of bytes. De-allocating
+     * can not transition us from the "used all" state back into the
+     * "used fewer bytes" state. It is a one way trip. Once we're in the
+     * "used all" state all de-allocates are done directly on the underlying
+     * breaker. So well behaved callers will naturally de-allocate everything.
+     * <p>
+     * {@link PreallocedCircuitBreaker#close()} is only used to de-allocate
+     * bytes from the underlying breaker if we're still in the "used fewer bytes"
+     * state. There is nothing to de-allocate if we are in the "used all" state. 
+     */
+    private static class PreallocedCircuitBreaker implements CircuitBreaker, Releasable {
+        private final CircuitBreaker next;
+        private final long preallocated;
+        private long preallocationUsed;
+        private boolean closed;
+
+        PreallocedCircuitBreaker(CircuitBreaker next, long preallocated) {
+            this.next = next;
+            this.preallocated = preallocated;
+        }
+
+        @Override
+        public void circuitBreak(String fieldName, long bytesNeeded) {
+            next.circuitBreak(fieldName, bytesNeeded);
+        }
+
+        @Override
+        public double addEstimateBytesAndMaybeBreak(long bytes, String label) throws CircuitBreakingException {
+            if (closed) {
+                throw new IllegalStateException("already closed");
+            }
+            if (preallocationUsed == preallocated) {
+                return next.addEstimateBytesAndMaybeBreak(bytes, label);
+            }
+            long newUsed = preallocationUsed + bytes;
+            if (newUsed > preallocated) {
+                preallocationUsed = preallocated;
+                long toAllocate = newUsed - preallocated;
+                if (toAllocate > 0) {
+                    return next.addEstimateBytesAndMaybeBreak(toAllocate, label);
+                }
+                return 0;
+            }
+            // This is the fast case. No volatile reads or writes here, ma!
+            preallocationUsed = newUsed;
+            // We return garbage here but callers never use the result for anything interesting
```

Comment:
Yeah. I feel like there are several "perspectives" on `CircuitBreaker` which might be useful to split up. The "I'm allocating and deallocating stuff" one. The "how much is used and how many times has this tripped?" one. There is kind of also a "testing" one. And that one is where the caller uses the result here. Maybe tests could just read the result rather than look at the percent. I'll take another look.